routers/user/auth.go

```diff
@@ -540,25 +540,43 @@ func SignInOAuthCallback(ctx *context.Context) {
 
 	u, gothUser, err := oAuth2UserLoginCallback(loginSource, ctx.Req.Request, ctx.Resp)
 
-	handleOAuth2SignIn(u, gothUser, ctx, err)
-}
-
-func handleOAuth2SignIn(u *models.User, gothUser goth.User, ctx *context.Context, err error) {
 	if err != nil {
 		ctx.ServerError("UserSignIn", err)
 		return
 	}
 
 	if u == nil {
-		// no existing user is found, request attach or new account
-		ctx.Session.Set("linkAccountGothUser", gothUser)
-		ctx.Redirect(setting.AppSubURL + "/user/link_account")
-		return
+		if setting.Service.EnableOAuth2AutoRegister {
+			// create new user with details from oauth2 provider
+			u = &models.User{
+				Name:        gothUser.UserID,
+				FullName:    gothUser.Name,
+				Email:       gothUser.Email,
+				IsActive:    !setting.Service.OAuth2RegisterEmailConfirm,
+				LoginType:   models.LoginOAuth2,
+				LoginSource: loginSource.ID,
+				LoginName:   gothUser.UserID,
```

Comment:
Might be good to check if `gothUser.NickName` contains only allowed gitea username characters and use that and only if it contains not allowed characters than fallback to UserID as in many providers userid will be cryptic uuid

Reply:
This sounds good. What should be the handling if two users have the same nickname attribute. I think that should be a configuration option. Depending on whiter the nickname attribute of the provider is unique or not. Or is the nickname unique in all common providers?

Reply:
Don't know about all, might be good if it would fallback to uid also if user with such nickname exists already

Reply:
+1